### PR TITLE
Closes #8122: Do not redirect to market app if app is already installed

### LINKED ISSUE #8122: Do not redirect to market app if app is already installed
Currently if app links failed to open the intent in an external app will fall back to the market intent.  This is with the assumption that if intent failed and target package exists, the app is not installed.

However, some site trying to redirect the intent to their app without correctly setting the intent filter on the app side.  If intent failed to launch and the package is already installed, use the fall back URL because something is wrong with the external app.

### PR DESCRIPTION
**Pull request by rocketsroger**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt [hunk 1]
```diff
@@ -173,6 +173,15 @@ class AppLinksUseCases(
...
         }
+
+        private fun isPackageInstalled(packageName: String): Boolean {
+            return try {
+                context.packageManager.getPackageInfo(packageName, PackageManager.GET_ACTIVITIES)
+                true
+            } catch (e: PackageManager.NameNotFoundException) {
+                false
+            }
+        }
```
Comment (Amejia481) [inline components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt:184]: Maybe can share this logic as an extension function as we are doing the same on other components [see](https://github.com/mozilla-mobile/android-components/blob/4202c70e488702cc9110b96997f5a5fc3860f7f7/components/support/utils/src/main/java/mozilla/components/support/utils/Browsers.kt#L232)

Comment (rocketsroger) [reply components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt:184]: Sure, I'll check if there's a good way to share this.  Thanks,

Comment (Amejia481) [reply components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt]: 👍

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/pm/PackageManager.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+package mozilla.components.support.ktx.android.content.pm
+
+import android.content.pm.PackageManager
+
+/**
+ * Check if a package is installed
+ *
+ * @param packageName The name of the package to check for.
+ */
+fun PackageManager.isPackageInstalled(packageName: String, flags: Int): Boolean {
```
Comment (Amejia481) [inline components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/pm/PackageManager.kt:14]: Nit: Can we add a test for this :)

File: components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt (hunk 1 above)
Comment (rocketsroger) [reply components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt]: Added the extension function.  Unfortunately `support-ktx` depends on `support-utils` causing a circular dependency so I couldn't replace the `Browsers.kt` call.

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/pm/PackageManager.kt (hunk 2 above)
Comment (rocketsroger) [reply components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/pm/PackageManager.kt:14]: Sure, I'll add it.  Thanks

**Comment by rocketsroger** [comment]
bors r=Amejia481

**Comment by silvanfischer** [comment]
@rocketsroger Is it possible that this change introduced the regression: https://github.com/mozilla-mobile/fenix/issues/14843?

**Comment by rocketsroger** [comment]
> @rocketsroger Is it possible that this change introduced the regression: [mozilla-mobile/fenix#14843](https://github.com/mozilla-mobile/fenix/issues/14843)?

I don't think so.  If anything it should help with the problem.  I'll self assign and look into the other issue when I have a chance.  Thanks

**Pull request merged**